cpp/src/io/parquet/error.hpp

```diff
@@ -50,30 +51,39 @@ class kernel_error {
    *
    * @param CUDA stream to use
    */
-  kernel_error(rmm::cuda_stream_view stream) : _error_code{0, stream} {}
+  kernel_error(rmm::cuda_stream_view stream) : _error_code(1, stream)
+  {
+    _error_code[0] = 0;
+    _error_code.host_to_device_async(stream);
+  }
 
   /**
    * @brief Return a pointer to the device memory for the error
    */
-  [[nodiscard]] auto data() { return _error_code.data(); }
+  [[nodiscard]] auto data() { return _error_code.device_ptr(); }
 
   /**
    * @brief Return the current value of the error
    *
-   * This uses the stream used to create this instance. This does a synchronize on the stream
-   * this object was instantiated with.
+   * @param stream The CUDA stream to synchronize with
    */
-  [[nodiscard]] auto value() const { return _error_code.value(_error_code.stream()); }
+  [[nodiscard]] auto value_sync(rmm::cuda_stream_view stream) const
+  {
+    _error_code.device_to_host_sync(stream);
+    return _error_code[0];
+  }
 
   /**
-   * @brief Return a hexadecimal string representation of the current error code
+   * @brief Return a hexadecimal string representation of an error code
    *
    * Returned string will have "0x" prepended.
+   *
+   * @param value The error code to convert to a string
    */
-  [[nodiscard]] std::string str() const
+  [[nodiscard]] static std::string to_string(value_type value)
   {
```

Comment:
Instead of static, shouldn't this be a free function instead of a member function?

Reply:
It could be, but I like how it reads at the call site: `kernel_error::to_string(error)` and it does not really change much else about the function.
I'm not insisting on this option, so, open to suggestions :)